clients/pkg/logentry/stages/logfmt_test.go

```diff
@@ -0,0 +1,312 @@
+package stages
+
+import (
+	"reflect"
+	"testing"
+	"time"
+
+	util_log "github.com/cortexproject/cortex/pkg/util/log"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stretchr/testify/assert"
+	"gopkg.in/yaml.v2"
+)
+
+var testLogfmtYamlSingleStageWithoutSource = `
+pipeline_stages:
+- logfmt:
+    mapping:
+      out:  message
+      app:
+      duration:
+      unknown:
+`
+
+var testLogfmtYamlMultiStageWithSource = `
+pipeline_stages:
+- logfmt:
+    mapping:
+      extra:
+- logfmt:
+    mapping:
+      user:
+    source: extra
```

Comment:
q: So if we have source, then entire `logfmt` parsing is only on that field correct? In this example, we try to parse all the mapping (only `user` here) from `extra` key on the log line.
^^ (this is super nice ❤️ )

How do we delimit the fields inside the source? (here its `=`. But can we have something like `extra="user:foo"` and still extract `user` mapping by having custom delimit `(":")`?

Reply:
There is no way to define a delimiter and logfmt has a standard format to use only `=`. I am using https://github.com/go-logfmt/logfmt for parsing the log message.